src/ReadWriteLocks.jl

```diff
@@ -1,42 +1,32 @@
 module ReadWriteLocks
 
-export ReadWriteLock, read_lock, write_lock, lock!, unlock!
-
-if isdefined(Base, :lock!)
-    import Base: lock!
-end
-
-if isdefined(Base, :unlock!)
-    import Base: unlock!
+using Base: lock, unlock
+if VERSION < v"1.2.0-"
+    using Base.Threads: AbstractLock
+else
+    using Base: AbstractLock
 end
 
-if !isdefined(Base, :Mutex)
-    typealias Mutex ReentrantLock
-
-    lock!(x::Mutex) = lock(x)
-    unlock!(x::Mutex) = unlock(x)
-end
-
-abstract _Lock
+export ReadWriteLock, read_lock, write_lock, lock!, unlock!
 
-immutable ReadLock{T<:_Lock}
+struct ReadLock{T<:AbstractLock}
     rwlock::T
 end
 
-immutable WriteLock{T<:_Lock}
+struct WriteLock{T<:AbstractLock}
     rwlock::T
 end
 
-type ReadWriteLock <: _Lock
+mutable struct ReadWriteLock <: AbstractLock
     readers::Int
     writer::Bool
-    lock::Mutex  # reentrant mutex
+    lock::ReentrantLock  # reentrant mutex
```

Comment:
I think we should make ReadWriteLock parameterized by an `AbstractLock` type, and just make `ReentrantLock` the default.

Reply:
one slight hiccup is we require v1.2 to have `ReentrantLock <: Base.AbstractLock`

Reply:
A `Union` is fine